Add buttons to help

TrustyJAID opened the pull request:
### Description of the changes

This adds support for button menus in help and adds a SimpleMenu cog creators can import to simply paginate items with buttons similar to the existing menu in terms of available features.

### Have the changes in this PR been tested?

<!--
Choose one (remove the line that doesn't apply):
-->
Yes
<!--
If the question doesn't apply (for example, it's not a code change), choose Yes.

Please respond to this question truthfully. We do not delay nor reject PRs
based on the answer to this question but it allows to better review this PR.
-->

TrustyJAID committed:
Add support for buttons in help command

TrustyJAID committed:
Add missing import

TrustyJAID committed:
I cannot type

TrustyJAID committed:
Document SimpleMenu

TrustyJAID committed:
black

TrustyJAID committed:
Behave the same way as menus.menu and don't display controls if only 1 page is present

TrustyJAID committed:
Add example usage, remove note about same type since it doesn't matter,
 add missing type that is technically possible

phenom4n4n commented on redbot/core/utils/views.py:
```diff
@@ -0,0 +1,174 @@
+        await interaction.response.edit_message(**kwargs)
+
+
+class SimpleMenu(discord.ui.View):
```
I think Views should be suffixed with View (`SimpleMenuView`? `PaginatedView`?) so it isn't confused with menus from the old ext library.

TrustyJAID replied:
I don't see any confusion here? SimpleMenu could be put into the utils.menus namespace but since it's a view I think it makes more sense inside the new utils.views namespace.

Kowlin replied:
I think SimpleMenu being placed here in views is totally fine and appropriate for its function and goal.

TrustyJAID committed:
Fix typehint on format_page

TrustyJAID committed:
Swap button argument order

TrustyJAID committed:
Cleanup typehints and docstrings

TrustyJAID committed:
Add missing HelpSetting

Jackenmen reviewed (commented):
- Buttons aren't used if the bot doesn't have react perms.
- `[p]helpset usebuttons` and `[p]helpset usemenus` are currently separate settings but it might make sense for the setting to work as a tribool. Either way, the user should be informed (in help, maybe also in command responses) that setting both doesn't work/is required for buttons to function (depending on what you choose to do here)

TrustyJAID commented:
> * Buttons aren't used if the bot doesn't have react perms.

Should we use buttons if menus are toggled and the bot does not have reaction perms despite buttons not being set then?

> * `[p]helpset usebuttons` and `[p]helpset usemenus` are currently separate settings but it might make sense for the setting to work as a tribool. Either way, the user should be informed (in help, maybe also in command responses) that setting both doesn't work/is required for buttons to function (depending on what you choose to do here)

I thought it was clear in the response "if menus are also enabled."

npc203 commented:
I'd support a tri-bool (a single command for toggling the three), one could get fancier by using a select menu for selecting the option perhaps.

TrustyJAID commented:
My concern with a tri-bool is that there will be 2 groups of people, some who want buttons and some who do not want buttons. With a tri-bool I now have to decide whether True uses buttons or False uses buttons which will then cause some people in either group to be forced to change their setting appropriately. The current way I have it setup the bot will startup the way they're expecting and can then optionally adjust their settings as they wish.

TrustyJAID committed:
Allow button help menu in channels the bot does not have reaction perms in

TwinDragon commented:
I'm with Trusty on this and having it as two separate settings. Though I do like buttons, for menus, I'd prefer a selection menu, but that's not what's being talked about. In this case, I think leaving it up to the bot owner to set whether or not the bot were to use buttons should remain as-is and would be easier from a UX point to help differentiate what they do, rather than having it all in a single setting.

TrustyJAID committed:
black

Jackenmen commented:
What I meant by tribool would be sth like `[p]helpset menus buttons/reactions/disable` (`[p]helpset menus` being the command)

TrustyJAID committed:
Clearly define all available types for button menus

TrustyJAID committed:
Make `[p]helpset usemenus` accept 3 options only instead of toggling, remove `[p]helpset usebuttons` in favor of new tribool approach.
Replace `help__use_buttons` in favor of `help__use_menus` being an enum instead potentially allowing further expansion and being more clear. Migrate existing `bool` values to `int` to utilize the new enum.

TrustyJAID committed:
Fix migration

TrustyJAID committed:
Fix help sending to DM's when menus are enabled

TrustyJAID committed:
black

TrustyJAID committed:
clarify help docstring

TrustyJAID committed:
Fix for zeph

TrustyJAID committed:
Remove the view on ephemeral messages just so they're clearly not interactable anymore

TrustyJAID committed:
use response.edit_message instead

TrustyJAID commented:
Something I have been thinking about is the possibility of attaching a select menu to the SimpleMenu. The idea being that if toggled the view would include the drop down selector showing up to 25 pages on the menu, if there's more than 25 the select menu would paginate e.g. if there's 50 pages and you're on page 25 it would show pages 13-37 on the drop down selector. Is this something people want included in this? Is this something people would want on `[p]help` as well? If it's included on `[p]help` how should it work? should it be included with all buttons menus or should there be a way to use just buttons with no drop down selector?

TrustyJAID committed:
Add select menu to SimpleMenu allowing easier navigation, include it on help for testing though might change depending on feedback

TrustyJAID committed:
black

TrustyJAID committed:
dict.update returns None

TrustyJAID committed:
Add bot owner to interaction check

TrustyJAID committed:
Make the select menu optional and allow the select menu to be the only thing sent on the SimpleMenu in addition to the stop button

TrustyJAID commented:
I have implemented my idea mentioned above and included options for buttons, buttons with select menu, and select menu only.

TrustyJAID committed:
Allow this menu to work better with hybrid commands

TrustyJAID committed:
Add stop button if there is only 1 page to view

TrustyJAID committed:
Allow only the author to interact with the menu to be consistent with menus

TrustyJAID committed:
Fix potential duplicate stop buttons being added

TrustyJAID committed:
Don't let SimpleMenu attempt to delete itself if it's sent as an ephemeral message

TrustyJAID committed:
Cleanup typehints, simplyify internals, Add `ephemeral` kwarg to start method that can only be used if the Context is from an interaction

TrustyJAID committed:
Cleanup `get_page`

TrustyJAID committed:
Add `disable_after_timeout` arg which disables the buttons instead of removing them

TrustyJAID committed:
update `[p]helpset usemenus` docstring

Zephyrkul commented on redbot/core/utils/views.py:
```diff
@@ -0,0 +1,258 @@
+        self.author = ctx.author
+        self.ctx = ctx
+        kwargs = await self.get_page(self.current_page)
+        self.message = await ctx.send(**kwargs, ephemeral=ephemeral)
```
I'm not sure if you plan on making this PR depend on a higher version of discord.py, but for the version of discord.py that Red currently installs this raises a `TypeError` for the unexpected `ephemeral=` kwarg.

TrustyJAID replied:
I expect that by the time this PR gets fully reviewed and merged we will have bumped our d.py version to one that includes this kwarg on ctx. I added this as a sort of future-proof in the event cog creators want to use the menu as an ephemeral response from a slash command.

TrustyJAID committed:
Fix `[p]helpset showsettings` for new enum method

Zephyrkul commented on redbot/core/core_commands.py:
```diff
@@ -3644,30 +3656,46 @@ async def helpset_resetsettings(self, ctx: commands.Context):
+            - `[p]helpset usemenus reactions` - Enables using reaction menus.
+            - `[p]helpset usemenus buttons` - Enables using button menus.
+            - `[p]helpset usemenus select` - Enables buttons with a select menu.
+            - `[p]helpset usemensus selectonly` - Enables a select menu only on help.
```
```suggestion
            - `[p]helpset usemenus selectonly` - Enables a select menu only on help.
```

TrustyJAID committed:
Update redbot/core/core_commands.py

Co-authored-by: Zephyrkul <23347632+Zephyrkul@users.noreply.github.com>

TrustyJAID committed:
Merge branch 'V3/develop' into add-help-buttons

TrustyJAID committed:
Fix style

TrustyJAID committed:
Merge branch 'V3/develop' into add-help-buttons

TrustyJAID committed:
Update doc strings to be more clear

TrustyJAID committed:
Fix passing list of dicts for pages not adding the view

TrustyJAID committed:
Improve docstrings and maybe make docs tests happier

TrustyJAID committed:
Block quote ends without a blank line; unexpected unindent. fix

Kowlin committed:
Added missing whitespace

Jackenmen committed:
Revert unrelated whitespace changes

Kowlin reviewed (approved):
Working like a charm, thank you.
